uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js

```diff
@@ -0,0 +1,224 @@
+
+      dialog.open({
+        title: 'Confirm ' + dialogParams.title + '?',
+        copyCancel: 'Cancel',
+        copyConfirm: dialogParams.btnLabel,
+        classConfirm: 'btn-' + dialogParams.btnClass,
+        msg: dialogParams.msg,
+        onConfirm: function () {
+          return vm.leaveRequest[action]()
+            .then(function () {
+              $rootScope.$emit('LeaveRequest::' + (action === 'delete' ? 'deleted' : 'edit'), vm.leaveRequest);
+            });
+        }
+      });
+    };
+
+    init();
+
+    function init () {
+      /* This component expects vm.leaveRequestStatuses and vm.absenceTypes
+       * to be an objects, so if arrays are passed, they are transformed into
+       * object with according indexes
+       */
+      if (Array.isArray(vm.leaveRequestStatuses)) {
+        vm.leaveRequestStatuses = _.indexBy(vm.leaveRequestStatuses, 'value');
+      }
+
+      if (Array.isArray(vm.absenceTypes)) {
+        vm.absenceTypes = _.indexBy(vm.absenceTypes, 'id');
+      }
+
+      var allowedActions = getAllowedActions();
+
+      vm.list = _.map(allowedActions, function (action) {
+        return {
+          key: action,
+          label: actions[action].label,
+          isDirectAction: actions[action].isDirectAction
+        };
+      });
+    }
+
+    /**
+     * @TODO This function utilises external resource
+     * vm.absenceTypes - this sould be refactored
+     *
+     * Checks if the given leave request can be cancelled
+     *
+     * Allow request cancellation values refer to the following constants:
+     * 1 = REQUEST_CANCELATION_NO
+     * 2 = REQUEST_CANCELATION_ALWAYS
+     * 3 = REQUEST_CANCELATION_IN_ADVANCE_OF_START_DATE
+     *
+     * @return {Boolean}
+     */
+    function canLeaveRequestBeCancelled () {
+      if (vm.role === 'admin' || vm.role === 'manager') {
+        return true;
+      }
+
+      var allowCancellationValue = vm.absenceTypes[vm.leaveRequest.type_id].allow_request_cancelation;
+
+      if (allowCancellationValue === '3') {
+        return moment().isBefore(vm.leaveRequest.from_date);
+      }
+
+      return allowCancellationValue === '2';
+    }
+
+    /**
+     * @TODO This function utilises external resource
+     * vm.leaveRequestStatuses - this sould be refactored
+     *
+     * Defines which actions can be taken with the
+     * leave request basing on its status, user role
+     *
+     * @return {Array} allowed actions
+     */
+    function getAllowedActions () {
+      var leaveRequestStatus = vm.leaveRequestStatuses[vm.leaveRequest.status_id].name;
+      var allowedActions = [];
+
+      _.each(actions, function (action, actionKey) {
+        if (_.include(action.statuses, leaveRequestStatus)) {
+          allowedActions.push(actionKey);
+        }
+      });
+
+      if (!canLeaveRequestBeCancelled()) {
+        _.pull(allowedActions, 'cancel');
+      }
+
+      if (vm.role !== 'admin') {
+        _.pull(allowedActions, 'delete');
+      }
```

Comment:
Just an idea to have them as one-liners and save space (i find it more readable also because of less white space but it's just my personal preference)
```js
!canLeaveRequestBeCancelled() && _.pull(allowedActions, 'cancel');
(vm.role !== 'admin') && _.pull(allowedActions, 'delete');
(vm.role === 'staff')  && _.pull(allowedActions, 'approve', 'reject');
```